File: chapters/algorithms/forward_euler_method/code/rust/euler.rs
fn main() {
    let mut result = [0.0; 100];
    let threshold = 0.01;
    let timestep = 0.01;

    solve_euler(timestep, &mut result);
    println!("{}", check_result(&result, threshold, timestep));
}

fn solve_euler(timestep: f64, result: &mut [f64]) {
    let n = result.len();
    if n != 0 {
        result[0] = 1.0;
        for i in 1..n {
            result[i] = result[i - 1] - 3.0 * result[i - 1] * timestep;
        }
    }
}

fn check_result(result: &[f64], threshold: f64, timestep: f64) -> bool {
    let mut is_approx: bool = true;
    for (i, val) in result.iter().enumerate() {
        let solution = (-3.0 * i as f64 * timestep).exp();
        if (val - solution).abs() > threshold {
            println!("{}    {}", val, solution);
            is_approx = false;
        }
    }

    return is_approx;
}

File: contents/forward_euler_method/code/cpp/euler.cpp
#include <algorithm>
#include <cmath>
#include <cstddef>
#include <iostream>
#include <iterator>
#include <utility>
#include <vector>

using std::begin;
using std::end;

using std::size_t;

std::vector<double> solve_euler(double timestep, size_t size) {
  std::vector<double> result;
  double current = 1.0;
  for (size_t i = 0; i < size; ++i) {
    result.push_back(current);
    current -= 3.0 * current * timestep;
  }
  return result;
}

// check_result takes an iterator over doubles,
// and returns whether any value is outside the passed threshold.
template <typename Iter>
bool check_result(Iter first, Iter last, double threshold, double timestep) {
  auto it = first;
  for (size_t idx = 0; it != last; ++idx, ++it) {
    double solution = std::exp(-3.0 * idx * timestep);
    if (std::abs(*it - solution) > threshold) {
      std::cout << "We found a value outside the threshold; the " << idx
                << "-th value was " << *it << ", but the expected solution was "
                << solution << '\n';
      std::cout << "(the threshold was " << threshold
                << " and the difference was " << std::abs(*it - solution)
                << ")\n";
      return true;
    }
  }
  return false;
}

int main() {
  double threshold = 0.01;
  double timestep = 0.01;

  auto result = solve_euler(timestep, 100);
  auto outside_threshold =
      check_result(begin(result), end(result), threshold, timestep);
  auto msg = outside_threshold ? "yes :(" : "no :D";

  std::cout << "Were any of the values outside of the threshold (" << threshold
            << ")? " << msg << '\n';
}
